conda_kapsel/archiver.py

```diff
@@ -329,3 +332,207 @@ def _archive_project(project, filename):
+                          (name, dest, canonical_project_dir))
+            return None
+        src_and_dest.append((name, dest))
+
+    return (canonical_project_dir, src_and_dest)
+
+
+class _UnarchiveStatus(SimpleStatus):
+    def __init__(self, success, description, logs, project_dir):
+        super(_UnarchiveStatus, self).__init__(success=success, description=description, logs=logs)
+        self.project_dir = project_dir
+
+
+# function exported for project_ops.py
+def _unarchive_project(archive_filename, project_dir, parent_dir=None):
+    """Unpack an archive of files in the project.
+
+    This takes care of several details, for example it deals with
+    hostile archives containing files outside of the dest
+    directory, and it handles both tar and zip.
+
+    It does not load or validate the unpacked project.
+
+    project_dir can be None to auto-choose one.
+
+    If parent_dir is non-None, place the project_dir in it. This is most useful
+    if project_dir is None.
+
+    Args:
+        archive_filename (str): the tar or zip archive file
+        project_dir (str): the directory that will contain the project config file
+        parent_dir (str): place project directory in here
+
+    Returns:
+        a ``Status``, if failed has ``errors``, on success has a ``project_dir`` property
+    """
+    if project_dir is not None and os.path.isabs(project_dir) and parent_dir is not None:
+        raise ValueError("If supplying parent_dir to unarchive, project_dir must be relative or None")
+
+    list_files = None
+    extract_files = None
+    if archive_filename.endswith(".zip"):
+        list_files = _list_files_zip
+        extract_files = _extract_files_zip
+    elif any([archive_filename.endswith(suffix) for suffix in [".tar", ".tar.gz", ".tar.bz2"]]):
+        list_files = _list_files_tar
+        extract_files = _extract_files_tar
+    else:
+        return SimpleStatus(
+            success=False,
+            description=("Could not unpack archive %s" % archive_filename),
+            errors=["Unsupported archive filename %s, must be a .zip, .tar.gz, or .tar.bz2" % (archive_filename)])
+
+    logs = []
+    errors = []
+    try:
+        result = _get_source_and_dest_files(archive_filename, list_files, project_dir, parent_dir, errors)
+        if result is None:
+            return SimpleStatus(success=False,
+                                description=("Could not unpack archive %s" % archive_filename),
+                                errors=errors)
+        (canonical_project_dir, src_and_dest) = result
+
+        if len(src_and_dest) == 0:
+            return SimpleStatus(success=False,
+                                description=("Could not unpack archive %s" % archive_filename),
+                                errors=["Archive does not contain a project directory or is empty."])
+
+        makedirs_ok_if_exists(canonical_project_dir)
```

Comment:
Since `_get_source_and_dest_files` has already verified that `canonical_project_dir` doesn't exist, it might be clearer to call `os.makedirs` here instead.